File: src/types/parameter.ts
export type ParameterType =
  | 'string'
  | 'integer'
  | 'float'
  | 'boolean'
  | 'enum'
  | 'vector2'
  | 'vector3'
  | 'vector4'
  | 'color'
  | 'file'
  | 'image'
  | 'expression'
  | 'array'
  | 'object'
  | 'json'
  | 'button'
  | 'separator'
  | 'label'
  | 'ramp'
  | 'keyframe'

export type PerformanceImpact = 'none' | 'low' | 'medium' | 'high' | 'critical'

export type ConditionOperator = 'eq' | 'ne' | 'gt' | 'gte' | 'lt' | 'lte' | 'contains' | 'not_contains'

export interface ParameterCondition {
  parameter: string
  operator: ConditionOperator
  value: unknown
}

export interface ParameterOption {
  value: unknown
  label: string
  icon?: string
}

export interface Parameter {
  id: string
  name: string
  displayName: string
  paramType: ParameterType
  defaultValue?: unknown
  minValue?: unknown
  maxValue?: unknown
  options?: ParameterOption[]
  description?: string
  group?: string
  visibleWhen?: ParameterCondition
  enabledWhen?: ParameterCondition
  performanceImpact: PerformanceImpact
  isAnimatable: boolean
  isExpressionCapable: boolean
  sortOrder: number
}

File: src-tauri/src/domain/parameter.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Parameter {
    pub id: Uuid,
    pub name: String,
    pub display_name: String,
    pub param_type: ParameterType,
    pub default_value: Option<serde_json::Value>,
    pub min_value: Option<serde_json::Value>,
    pub max_value: Option<serde_json::Value>,
    pub options: Option<Vec<ParameterOption>>,
    pub description: Option<String>,
    pub group: Option<String>,
    pub visible_when: Option<ParameterCondition>,
    pub enabled_when: Option<ParameterCondition>,
    pub performance_impact: PerformanceImpact,
    pub is_animatable: bool,
    pub is_expression_capable: bool,
    pub sort_order: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum ParameterType {
    String,
    Integer,
    Float,
    Boolean,
    Enum,
    Vector2,
    Vector3,
    Vector4,
    Color,
    File,
    Image,
    Expression,
    Array,
    Object,
    Json,
    Button,
    Separator,
    Label,
    Ramp,
    Keyframe,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParameterOption {
    pub value: serde_json::Value,
    pub label: String,
    pub icon: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParameterCondition {
    pub parameter: String,
    pub operator: ConditionOperator,
    pub value: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ConditionOperator {
    Eq,
    Ne,
    Gt,
    Gte,
    Lt,
    Lte,
    Contains,
    NotContains,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "snake_case")]
pub enum PerformanceImpact {
    #[default]
    None,
    Low,
    Medium,
    High,
    Critical,
}
